### PR TITLE
refactor(amplify-category-hosting): hosting ux improvement

### PR DESCRIPTION
**Pull request by UnleashedMind**
*Issue #, if available:*
#261
#593
*Description of changes:*
- index and error doc configuration input validation
- sanity check before uploading files, warn user if index file is missing, error out if distribution folder is missing or empty
- If hosting bucket has not been created yet (in between `add` and a successful `push`), `configure` allows the user to change the bucket name. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-hosting/lib/S3AndCloudFront/configuration-manager.js [hunk 1]
```diff
@@ -8,11 +8,68 @@ const configurables = {
...
+  const bucketCreated = serviceMeta && serviceMeta.output && serviceMeta.output.HostingBucketName;
+  if (!bucketCreated) {
+    const bucketName = context.exeInfo.template.Resources.S3Bucket.Properties.BucketName;
+    await setBucketName(context, bucketName);
+  }
+}
+
+async function configureHostingComponents(context, lastConfiguredSection) {
+  const options = Object.keys(configurables);
+  const done = "I'm done.";
```
Comment (kaustavghosh06) [inline packages/amplify-category-hosting/lib/S3AndCloudFront/configuration-manager.js]: Can we change this text to something more meaningful?

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js [hunk 2]
```diff
@@ -8,22 +8,39 @@ async function configure(context) {
...
 
   return context;
 }
 
+function docKeyValidation(str) {
+  str = str.trim();
+
+  let isValid = str.length > 0;
+  if (!isValid) {
+    return 'Must not be empty, or only contains space charactors.';
```
Comment (kaustavghosh06) [inline packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js]: What should not be empty?

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js [hunk 3]
```diff
@@ -8,22 +8,39 @@ async function configure(context) {
...
+  }
+
+  isValid = !/\//.test(str);
+  if (!isValid) {
+    return 'The slash charactor is not allowed.';
```
Comment (kaustavghosh06) [inline packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js]: Is backslash allowed?

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-scanner.js [hunk 4]
```diff
@@ -2,11 +2,36 @@ const fs = require('fs-extra');
...
+      context.print.info('');
+      throw new Error(message);
+    } else if (!fs.existsSync(path.join(distributionDirPath, indexDoc))) {
+      context.print.info('');
+      context.print.warning('Index doc is missing in the distribution folder');
+      context.print.info('Distribution folder is currently set as:');
+      context.print.info(`  ${distributionDirPath}`);
+      context.print.info('Index document is currently set as:');
+      context.print.info(`  ${indexDoc}`);
+      context.print.info('');
```
Comment (kaustavghosh06) [inline packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-scanner.js]: No error is thrown in this case?

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-uploader.js [hunk 5]
```diff
@@ -9,20 +10,27 @@ const serviceName = 'S3AndCloudFront';
...
     uploadFileTasks.push(() =>
       uploadFile(s3Client, hostingBucketName, distributionDirPath, filePath));
   });
 
-  return sequential(uploadFileTasks);
+  const spinner = new Ora('Uploading files...');
+  try {
+    spinner.start();
+    await sequential(uploadFileTasks);
+    spinner.succeed('Uploading files successful.');
```
Comment (kaustavghosh06) [inline packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-uploader.js]: Uploaded files successfully.

File: packages/amplify-category-hosting/lib/S3AndCloudFront/configuration-manager.js (hunk 1 above)
Comment (UnleashedMind) [reply packages/amplify-category-hosting/lib/S3AndCloudFront/configuration-manager.js]: any suggestions?

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js (hunk 3 above)
Comment (UnleashedMind) [reply packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/configure-Website.js]: backslash is actually allowed, although probably very ugly.

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-scanner.js (hunk 4 above)
Comment (UnleashedMind) [reply packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-scanner.js]: no need to throw error in this case.

File: packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-uploader.js (hunk 5 above)
Comment (UnleashedMind) [reply packages/amplify-category-hosting/lib/S3AndCloudFront/helpers/file-uploader.js]: check

**Pull request merged**